### DISCUSSION TITLE
How to enable auto redirect when I want to use only domain based routing?

### DISCUSSION THREAD

**Discussion by siamahnaf** [amannn/next-intl#1159] [General]
Hello, I want to create a nextjs i18 app with subdomain based locale, like-

`example.com //It is default locale, on default locale I want to use en`
`fr.example.com`
`de.example.com`

I don't want to use prefix like-
`fr.example.com/fr`
`de.example.com/de`

That's why disable the locale prefix. Following is my code-

`i18n.ts`-

```
import { notFound } from "next/navigation";
import { getRequestConfig } from "next-intl/server";

const locales = ["en", "fr", "de"];

export default getRequestConfig(async ({ locale }) => {
    if (!locales.includes(locale as any)) notFound();
    return {
        messages: (await import(`../translate/${locale}.json`)).default
    };
});
```

`middleware.ts`-

```
import createMiddleware from "next-intl/middleware";

export default createMiddleware({
    locales: ["en", "fr", "de"],
    defaultLocale: "en",
    localePrefix: "never",
    domains: [
        {
            domain: "dev.localhost:3000",
            locales: ["en"],
            defaultLocale: "en"
        },
        {
            domain: "fr.dev.localhost:3000",
            locales: ["fr"],
            defaultLocale: "fr"
        },
        {
            domain: "de.dev.localhost:3000",
            locales: ["de"],
            defaultLocale: "de"
        }
    ]
});

export const config = {
    matcher: ['/((?!api|_next|.*\\..*).*)']
};
```

So, here you can see that, I use this `localePrefix: "never",`. Now it is not auto redirecting.

What I want?
I want to auto redirect to subdomain based routing without adding locale prefix. Suppose, detection locale is fr, but I am going to root domain `example.com`, it automatically should redirect  to `fr.example.com`.

Again suppose, detection locale is en, but I am trying to go `fr.example.com` it should redirect to `example.com`, as here I want to use root domain for en locale.

Same as, if I previously switch locale to fr, next-intl is saving this onto cookie, then I try to go `example.com`, it should redirect to `fr.example.com`.

Could anyone please help me, I need to understand the whole things. Please say or give minimal example, how can I do that?

**Comment by amannn** [comment]
`next-intl` will use the `defaultLocale` of your domain as a fallback in your case, see [the docs](https://next-intl-docs.vercel.app/docs/routing/middleware#location-detection-domain):

> (4) As a fallback, the `defaultLocale` of the domain is used

Therefore no automatic redirect is done that would change the domain.

You can however [compose the middleware](https://next-intl-docs.vercel.app/docs/routing/middleware#composing-other-middlewares) and add additional redirects on top if you like before the `next-intl` middleware handles the request.

**Comment by Matt-Cain** [comment]
Hi 👋 , I am also running into a similar issue with auto redirects to subdomains not working as expected.

Expected:
- When I navigate to `local.app.com` with `Accept-Language: pt-BR;q=0.9`
- I should be redirect to `pt.local.app.com`

Actually:
- When I navigate to `local.app.com` with `Accept-Language: pt-BR;q=0.9`
- I am not redirected, but stay on `local.app.com`
- However, the content on the page shows the correct language

```.ts
import { defineRouting } from 'next-intl/routing';

export const routing = defineRouting({
  defaultLocale: 'en-US',
  domains: [
    {
      defaultLocale: 'en-US',
      domain: 'en.local.app.com',
      locales: ['en-US'],
    },
    {
      defaultLocale: 'pt-BR',
      domain: 'pt.local.app.com',
      locales: ['pt-BR'],
    },
    {
      defaultLocale: 'es',
      domain: 'es.local.app.com',
      locales: ['es'],
    },
  ],
  localePrefix: { mode: 'never' },
  locales: ['en-US', 'pt-BR', 'es'],
});
```

Important to note: When I set localePrefix to `{ mode: 'always' }`, I get redirected properly, albeit prefix-based routing

```
    "next": "15.3.1",
    "next-intl": "^4.1.0",
    "react": "^19.0.0",
```
